x-pack/legacy/plugins/monitoring/public/lib/setup_mode.js

```diff
@@ -111,70 +116,64 @@ export const disableElasticsearchInternalCollection = async () => {
 };
 
 export const toggleSetupMode = inSetupMode => {
-  return new Promise(async (resolve, reject) => {
-    try {
-      checkAngularState();
-    } catch (err) {
-      return reject(err);
-    }
-
-    const globalState = angularState.injector.get('globalState');
-    setupModeState.enabled = inSetupMode;
-    globalState.inSetupMode = inSetupMode;
-    globalState.save();
-    setSetupModeMenuItem(); // eslint-disable-line  no-use-before-define
-    notifySetupModeDataChange();
-
-    if (inSetupMode) {
-      await updateSetupModeData();
-    }
+  checkAngularState();
 
-    resolve();
-  });
+  const globalState = angularState.injector.get('globalState');
+  setupModeState.enabled = inSetupMode;
+  globalState.inSetupMode = inSetupMode;
+  globalState.save();
+  setSetupModeMenuItem(); // eslint-disable-line  no-use-before-define
+  notifySetupModeDataChange();
+
+  if (inSetupMode) {
+    // Intentionally do not await this so we don't block UI operations
+    updateSetupModeData();
+  }
 };
 
 const setSetupModeMenuItem = () => {
-  // Disabling this for this initial release. This will be added back in
-  // in a subsequent PR
-  // checkAngularState();
-
-  // const globalState = angularState.injector.get('globalState');
-  // const navItems = globalState.inSetupMode
-  //   ? [
-  //     {
-  //       key: 'exit',
-  //       label: 'Exit Setup Mode',
-  //       description: 'Exit setup mode',
-  //       run: () => toggleSetupMode(false),
-  //       testId: 'exitSetupMode'
-  //     },
-  //     {
-  //       key: 'refresh',
-  //       label: 'Refresh Setup Data',
-  //       description: 'Refresh data used for setup mode',
-  //       run: () => updateSetupModeData(),
-  //       testId: 'refreshSetupModeData'
-  //     }
-  //   ]
-  //   : [{
-  //     key: 'enter',
-  //     label: 'Enter Setup Mode',
-  //     description: 'Enter setup mode',
-  //     run: () => toggleSetupMode(true),
-  //     testId: 'enterSetupMode'
-  //   }];
-
-  // angularState.scope.topNavMenu = [...navItems];
+  checkAngularState();
+
+  if (isOnPage('no-data')) {
+    return;
+  }
+
+  const globalState = angularState.injector.get('globalState');
+  const navItems = globalState.inSetupMode
+    ? []
+    : [{
+      id: 'enter',
+      label: 'Enter Setup Mode',
+      description: 'Enter setup',
+      run: () => toggleSetupMode(true),
+      testId: 'enterSetupMode'
+    }];
+
+  angularState.scope.topNavMenu = [...navItems];
+  // LOL angular
+  if (!angularState.scope.$$phase) {
+    angularState.scope.$apply();
+  }
```

Comment:
I think here you can get away with something like:
```javascript
angularState.scope.$apply(() => angularState.scope.topNavMenu = [...navItems])
```

Otherwise you can actually miss apply if phase is in digest mode